core/src/allocation.jl

```diff
@@ -0,0 +1,503 @@
+"""Whether the given node node is flow constraining by having a maximum flow rate."""
+is_flow_constraining(node::AbstractParameterNode) = hasfield(typeof(node), :max_flow_rate)
+
+"""Whether the given node is flow direction constraining (only in direction of edges)."""
+is_flow_direction_constraining(node::AbstractParameterNode) =
+    (nameof(typeof(node)) ∈ [:Pump, :Outlet, :TabulatedRatingCurve])
+
+"""Find out whether a path exists between a start node and end node in the given graph."""
+function path_exists(graph::DiGraph, start_node_id::Int, end_node_id::Int)::Bool
+    node_ids_visited = Set{Int}()
+    stack = [start_node_id]
+
+    while !isempty(stack)
+        current_node_id = pop!(stack)
+        if current_node_id == end_node_id
+            return true
+        end
+        if !(current_node_id in node_ids_visited)
+            push!(node_ids_visited, current_node_id)
+            for outneighbor_node_id in outneighbors(graph, current_node_id)
+                push!(stack, outneighbor_node_id)
+            end
+        end
+    end
+    return false
+end
+
+"""
+Construct the JuMP.jl model for allocation.
+
+Definitions
+-----------
+- 'subnetwork' is used to refer to the original Ribasim subnetwork;
+- 'AG' is used to refer to the allocation graph.
```

Comment:
I'd stick to the lowercase conventions for variable and function names: `ag`, just like we talk about IDs in docs but `ids` in code.
`ag` is a bit cryptic so it's good you define it. It sounds more like agriculture to me.
Allocation is commonly abbreviated to alloc. `allocgraph` is not too bad.